src/monitor.rs

```diff
@@ -33,33 +33,49 @@ impl AppState {
 
     /// Poll the Ethereum node for the latest block and update the shared state
     /// with the latest block.
-    #[tracing::instrument(skip(app_state))]
+    #[tracing::instrument(skip_all)]
     pub(crate) async fn poll_and_update_block<
         T: Transport + Clone,
         P: Provider<T, N> + std::fmt::Debug,
         N: Network,
     >(
+        block: Block,
         provider: P,
         app_state: Arc<Mutex<AppState>>,
     ) {
-        let block = provider
-            .get_block_by_number(BlockNumberOrTag::Latest, true)
-            .await;
-        let mut state = app_state.lock().unwrap();
-        match block {
-            Ok(Some(block)) => {
-                state.latest_block = Some(block);
-                let block = state.latest_block.as_ref().unwrap();
-                tracing::debug!(
-                    block.number = block.header.number.map(|num| num.to::<u64>()),
-                    block.timestamp = block.header.timestamp.to::<u64>(),
-                    block.hash = block.header.hash.unwrap_or_default().to_string(),
-                    "Updated latest block"
-                );
+        // Before updating the latest block, check to see if it is retrievable from the
+        // HTTP provider. If the block is not found in the HTTP provider, log an error
+        // and continue to the next block.
+        // If the block is found in the HTTP provider, check to see if the block hash
+        // matches between the HTTP and WS providers. If the block hash does not match,
+        // log an error and continue to the next block.
+        let http_block = provider
+            .get_block_by_number(
+                BlockNumberOrTag::Number(block.header.number.unwrap().to::<u64>()),
+                true,
+            )
+            .await
+            .unwrap();
+
+        match http_block {
+            Some(blk) => {
+                if block.header.hash != blk.header.hash {
+                    tracing::error!("Block hash mismatch between HTTP and WS providers");
+                }
+            }
+            None => {
+                tracing::error!("Block not found in HTTP provider");
             }
-            Ok(None) => tracing::error!("No block found"),
-            Err(e) => tracing::error!("Failed to get latest block: {:?}", e),
         }
+
+        tracing::debug!(
+            block.number = block.header.number.map(|num| num.to::<u64>()),
+            block.timestamp = block.header.timestamp.to::<u64>(),
+            block.hash = block.header.hash.unwrap_or_default().to_string(),
+            "Updating to latest block"
+        );
+        let mut state = app_state.lock().unwrap();
+        state.latest_block = Some(block);
```

Comment:
Nit: this should work and does not present the opportunity to add some slow code at the end of the function while holding the lock.
```suggestion
        app_state.lock().unwrap().latest_block = Some(block);
```